rust/src/core.rs

```diff
@@ -227,6 +229,55 @@ pub(crate) fn verify_kernel_hmac(rootfs: i32, moddir: &str) -> CxxResult<()> {
     verify_kernel_hmac_impl(&moddir).map_err(Into::into)
 }
 
+pub(crate) fn stage_container_rpms(rpms: Vec<String>) -> CxxResult<Vec<String>> {
+    let rpms: Result<Vec<File>> = rpms
+        .into_iter()
+        .map(|path| File::open(path).map_err(|e| e.into()))
+        .collect();
+    stage_container_rpm_files(rpms?)
+}
+
+pub(crate) fn stage_container_rpm_raw_fds(fds: Vec<i32>) -> CxxResult<Vec<String>> {
+    stage_container_rpm_files(
+        fds.into_iter()
+            .map(|fd| unsafe { File::from_raw_fd(fd) })
+            .collect(),
+    )
+}
+
+fn stage_container_rpm_files(rpms: Vec<File>) -> CxxResult<Vec<String>> {
+    let mut r = Vec::new();
+    let mut sack = dnf_sack_new();
+    // XXX: This is really ugly: libdnf enforces that the filename ends in `.rpm`. So we use this
+    // tempdir to hold symlinks to the fdpaths to fool it. Yuck. And we can't use cap_tempfile here
+    // because the symlinks we create lead outside.
+    let d = tempfile::tempdir()?;
+    for mut rpm in rpms.into_iter() {
+        let fdpath = format!("/proc/self/fd/{}", rpm.as_raw_fd());
+        let symlink = format!("{}/{}.rpm", d.path().to_str().unwrap(), rpm.as_raw_fd());
+        std::os::unix::fs::symlink(&fdpath, &symlink)?;
+        let mut pkg = sack.pin_mut().add_cmdline_package(symlink)?;
+        let chksum = crate::ffi::get_repodata_chksum_repr(&mut pkg.pin_mut().get_ref())?;
+        if !chksum.starts_with("sha256:") {
+            return Err(anyhow!(
+                "Expected SHA256 hash, got {}. Please report this as a bug.",
+                &chksum
+            )
+            .into());
+        }
```

Comment:
Ahh nice, hadn't encountered `anyhow::ensure!` before.

Reply:
OK sadly, this doesn't work in this particular instance because of the `CxxResult` mismatch. The `split_once()` suggestion is still good and applicable though.